testapp/testapp/AppDelegate.swift

```diff
@@ -0,0 +1,156 @@
+        let options = ARTClientOptions()
+        options.authCallback = { params, callback in
+            self.getTokenRequest() { json, error in
+                do {
+                    callback(try ARTTokenRequest.fromJson(json!), nil)
+                } catch let error as NSError {
+                    callback(nil, error)
+                }
+            }
+        }
+        realtime = ARTRealtime(options: options)
+        realtime.connection.on { state in
+            if let state = state {
+                switch state.current {
+                case .connected:
+                    print("connected")
+                case .failed:
+                    print("failed")
+                default:
+                    break
+                }
+            }
+        }
+        return realtime
+    }
+
+
+    // 4
+    func didActivateAblyPush(_ error: ARTErrorInfo?) {
+        if let error = error {
+            // Handle error
+            print("** push activation failed, err=\(String(describing: error))")
+            return
+        }
+        print("** push activated")
+
+        self.channel = self.realtime.channels.get("push")
+        self.channel.attach() { (err) in
+            print("** channel attached, err=\(String(describing: err))")
+
+            self.channel.push.subscribeDevice { (err) in
+                DispatchQueue.main.async {
+                    print("** device ID \(self.realtime.device.id)")
+                }
+                print("** channel.push.subscribeDevice: err=\(String(describing: err))")
+                self.subscribed = true
+            }
+        }
+    }
+
+    // 5
+    func didDeactivateAblyPush(_ error: ARTErrorInfo?) {
+        if let error = error {
+            // Handle error
+            print("** push de-activation failed", error)
+            return
+        }
+        print("** push de-activated, re-activating")
+        self.realtime.push.activate()
+    }
+
+    // 6
+    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
+        print("** received notification: \(userInfo)")
+    }
+
+    // 7
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        // Override point for customization after application launch.
+        print("** hello")
+
+        UNUserNotificationCenter.current().requestAuthorization(options:[.badge, .alert, .sound]) { (granted, err) in
+            DispatchQueue.main.async() {
+                UIApplication.shared.registerForRemoteNotifications()
+                print("** after registerForRemoteNotifications")
+            }
+        }
+
+        self.realtime = self.getAblyRealtime()
+
+        // Deactivating may be necessary as the deployed app might have state inconsistent with the Ably app
+        // print("** De-activating Ably push")
+        // self.realtime.push.deactivate()
```

Comment:
@cjbatin Didn't got this one, sorry. Could you clarify?

Reply:
This was from the sample app I was provided